src/huggingface_hub/hf_file_system.py

```diff
@@ -0,0 +1,432 @@
+import itertools
+import os
+import tempfile
+from dataclasses import dataclass
+from datetime import datetime
+from glob import has_magic
+from typing import Any, Dict, List, Optional, Tuple, Union
+from urllib.parse import quote, unquote
+
+import fsspec
+import requests
+
+from ._commit_api import CommitOperationDelete
+from .constants import DEFAULT_REVISION, ENDPOINT, REPO_TYPE_MODEL, REPO_TYPES_MAPPING, REPO_TYPES_URL_PREFIXES
+from .hf_api import HfApi
+from .utils import (
+    EntryNotFoundError,
+    HFValidationError,
+    RepositoryNotFoundError,
+    RevisionNotFoundError,
+    hf_raise_for_status,
+    http_backoff,
+    paginate,
+    parse_datetime,
+)
+
+
+@dataclass
+class HfFileSystemResolvedPath:
+    """Data structure containing information about a resolved hffs path."""
+
+    repo_type: str
+    repo_id: str
+    revision: str
+    path_in_repo: str
+
+    def unresolve(self):
+        path = (
+            f"{REPO_TYPES_URL_PREFIXES.get(self.repo_type, '') + self.repo_id}@{self.revision}/{self.path_in_repo}"
```

Comment:
I found this unresolve() returns unqoted revisions, e.g. `'datasets/lhoestq/demo1@refs/convert/parquet'`

this might fix it ?
```suggestion
            f"{REPO_TYPES_URL_PREFIXES.get(self.repo_type, '') + self.repo_id}@{quote(self.revision, safe='')}/{self.path_in_repo}"
```

Reply:
Both `ls` commands work with the fix you suggested

```py
from huggingface_hub import HfFileSystem

fs = HfFileSystem()

fs.ls("datasets/lhoestq/demo1", revision="refs/convert/parquet")
fs.ls("datasets/lhoestq/demo1@refs%2Fconvert%2Fparquet")
```

Reply:
I can confirm it works:
```python
>>> fs.ls("datasets/lhoestq/demo1", revision="refs/convert/parquet")
[{'name': 'datasets/lhoestq/demo1/lhoestq--demo1',
  'size': 0,
  'type': 'directory'}]
```
```python
>>> fs.ls("datasets/lhoestq/demo1@refs%2Fconvert%2Fparquet")
[{'name': 'datasets/lhoestq/demo1@refs%2Fconvert%2Fparquet/lhoestq--demo1',
  'size': 0,
  'type': 'directory'}]
```